extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java

```diff
@@ -900,33 +905,52 @@ private static void setHttpServerTiming(boolean httpDisabled, HttpServerOptions
 
         if (!httpDisabled && httpServerOptions != null) {
             serverListeningMessage.append(String.format(
-                    "http://%s:%s", httpServerOptions.getHost(), actualHttpPort));
+                    "http://%s:%s", getDeveloperFriendlyHostName(httpServerOptions), actualHttpPort));
             socketCount++;
         }
 
         if (sslConfig != null) {
             if (socketCount > 0) {
                 serverListeningMessage.append(" and ");
             }
-            serverListeningMessage.append(String.format("https://%s:%s", sslConfig.getHost(), actualHttpsPort));
+            serverListeningMessage
+                    .append(String.format("https://%s:%s", getDeveloperFriendlyHostName(sslConfig), actualHttpsPort));
             socketCount++;
         }
 
         if (domainSocketOptions != null) {
             if (socketCount > 0) {
                 serverListeningMessage.append(" and ");
             }
-            serverListeningMessage.append(String.format("unix:%s", domainSocketOptions.getHost()));
+            serverListeningMessage.append(String.format("unix:%s", getDeveloperFriendlyHostName(domainSocketOptions)));
         }
         if (managementConfig != null) {
             serverListeningMessage.append(
                     String.format(". Management interface listening on http%s://%s:%s.", managementConfig.isSsl() ? "s" : "",
-                            managementConfig.getHost(), actualManagementPort));
+                            getDeveloperFriendlyHostName(managementConfig), actualManagementPort));
         }
 
         Timing.setHttpServer(serverListeningMessage.toString(), auxiliaryApplication);
     }
 
+    /**
+     * To improve developer experience in WSL dev/test mode, the server listening message should print "localhost" when
+     * the host is set to "0.0.0.0". Otherwise, display the actual host.
+     * Do not use this during the actual configuration, use options.getHost() there directly instead.
+     */
+    private static String getDeveloperFriendlyHostName(HttpServerOptions options) {
+        return (isWSL() && LaunchMode.current().isDevOrTest() && options.getHost().equals("0.0.0.0")) ? "localhost"
```

Comment:
I don't think `getHost()` can ever return `null` but I would use the safe pattern here too and have `0.0.0.0` first.